### DISCUSSION TITLE
Multiple resolutions

### DISCUSSION THREAD

**Discussion by Jauntiness** [itsToggle/plex_debrid#59] [Ideas]
Hi

is it possible to set your tool to load different resolutions if available, e.g. 2160p HDR + 1080p, so in Plex I could choose the resolution that fits best to my current device?

(Btw, I am very thankful for this tool. It was defnitley worth setting it up and testing, your Description is very helpful. Everything seems to work very good.  I am using "RaiDrive" since "rclone" wasnt working for me)

**Comment by itsToggle** [comment]
That's a great idea!

**Comment by itsToggle** [comment]
Hey, Im currently finishing up some bug fixes and wanted to start with this implementation. How would you like this to work? What Im thinking right now would be one of these two options:

1) You define a single trigger phrase (regex definition). If a downloaded release matches this trigger, a second release that **does not** match this trigger will be downloaded aswell. So if you enter "(HDR)", the script will try to download an HDR release aswell as the next best release thats not HDR. Another example would be entering "(2160)", which will cause the script to download a 2160p release and the next best release that is not 2160p.

2) The second option would be that you explicitly define what should be downloaded by multiple trigger phrases. If you enter "(2160)|(1080)|(720)" the script will try to download the best possible release for all specified qualities. This has the benefit of more control over what is downloaded, but I dont know how I could write this in a way that allows downloading HDR and non-HDR versions. Thats because most non-HDR releases dont have "SDR" or something similar in their name, so you cant explicitly look for a match by entering "(HDR)|(SDR)".

If anyone has an idea on how these two options could be combined in a single setting, Id be very thankful :)

**Comment by Jauntiness** [comment]
Hey,

I am amateur but I like to share my thoughts since I wished this option ;) 
The second option sounds defnitley more elegant to me, but I see it can get messy.

Cant you search for releases that do not contain "HDR"? It should contain 2160 but not HDR. E.g. something like this  "(2160)|(2160) NOT (HDR)|(1080)|(720)"

Otherwise you already have the option "release exclusion:" . Couldnt you trigger one search for ascending and one for descending - so one time releases with "(\.HDR\.)" get preffered and the next time releases without are. 

Cheers,

**Reply by itsToggle** [reply]
Yeah I'm an idiot haha. I was attempting to make this possible with a single regex string, but I think I can just make the setting have multiple "rules", similar to the "release sorting" rules. The ascending/descending switch, as you suggested, is a great idea!

**Reply by Jauntiness** [reply]
Great if I could help! :)

**Comment by itsToggle** [comment]
Latest version of plex_debrid now allows you to specify multiple version definitions :)

Check the readmes "Usage" section for some examples.

plex_debrid will first sort the scraped releases by your sorting rules and then download the best release for each version you specify.

**Reply by Jauntiness** [reply]
Nice! Thank you for the work!
I did test the new settings a bit but it doesnt alway work as intended. Unfortunatley I did not find any "logic" why and when it doesnt work. 

E.g. When I set more than two multiple versions it usually ignores at least one. Three rules with "(720p)" "(1080p)" and "(2160p)" might only provide a 720p and a 2160p result.


Right now I use "(1080p)" and "(2160p)" - that does seem to work stable and provides those two versions, usually a 1080 non HDR and a 2160 HDR version. Exactly what I wished for!

When I try "`(\.HDR\.)`" I get an error message I dont understand:

> [2022-08-20 13:53:02.823015] attempting to download version 4/4: "`(\.HDR\.)`" ... Exception in thread Thread-1201 (run):
Traceback (most recent call last):
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\Scripts\plex_rd.py", line 3039, in run
    element.download(library=library)
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\Scripts\plex_rd.py", line 360, in download
    if self.debrid_download():
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\Scripts\plex_rd.py", line 456, in debrid_download
    if debrid.download(self,stream=True):
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\Scripts\plex_rd.py", line 1710, in download
    if service.download(element,stream=stream,query=query,force=force):
  File "C:\Users\*\AppData\Local\Programs\Python\Python310\Scripts\plex_rd.py", line 1863, in download
    if len(version.files) > 0 and version.wanted > len(wanted)/2 or force:
AttributeError: 'file' object has no attribute 'files' 
>


But since the two resolutions seem to already work I am happy ;)

**Reply by itsToggle** [reply]
That exception is something I should fix, I'll look into it. edit: fixed in latest commit, was caused by downloading the same release twice.

plex_debrid will first and foremost try to only download releases that are cached (instantly available) on the debrid service/s of your choice. Sometimes, there simply are no cached releases of a quality/version you are looking for. If this happens, the current procedure is to simply accept that some versions couldn't be found. As long as at least one cached release was added, plex_debrid will deem the content as downloaded. The alternative here is to add an uncached release, which will take several minutes or hours to download.. 

If no cached release was found at all (and uncached releases are enabled, which they are by default), plex_debrid will add an uncached release for each version that you specify. This too is limited by the fact that sometimes there simply are no torrents for a given version.

You can check which releases plex_debrid will be able to add by simply scraping manually from the main menu scraper. There is a row called "cached" which indicates on which debrid service/s a given release is cached.

You can increase the chances of finding a cached release by integrating jackett and adding a decent amount of torrent indexers to it.

**Reply by Jauntiness** [reply]
Okay, thank you for your explanation. That makes sense and will likely explain why sometimes a version wasnt added! I will try to manually scrape and compare.

**Reply by Jauntiness** [reply]
I tried setting up another device (Plex, Trakt, Jackett, RealDebrid). Sorting rules are for (2160|1080|720) and file size is descending.
Unfortunatley on this device the multpile version setting isnt working at all. I add the log, maybe that helps understanding if it is a user error or something else.

> *Type 'exit' to return to the main menu.
[2022-08-20 21:09:59.896233] [plex] getting entire plex library ... done
[2022-08-20 21:10:00.005787] [plex] getting all watchlists ... done
[2022-08-20 21:10:17.392446] checking new content ... done
[2022-08-20 21:10:23.349587] [plex] item: "Death on the Nile" found in Plex-Online-Librarys watchlist
[2022-08-20 21:10:23.663580] [plex] getting entire plex library ... done
[2022-08-20 21:10:23.762622] checking new content ... done
[2022-08-20 21:10:23.766571] scraping sources for query "death.on.the.nile.2022" ... done - found 150 releases
[2022-08-20 21:10:27.403528] attempting to download version 1/4: "(1080p) " ... done
[2022-08-20 21:10:27.408106] attempting to download version 2/4: "(2160p)" ... done
[2022-08-20 21:10:28.661906] [realdebrid] adding cached release: Death.on.the.Nile.2022.2160p.Bluray.TrueHD.7.1.Atmos.HDR.x265-EVO... done
[2022-08-20 21:10:28.663753] attempting to download version 3/4: "(720p)" ... done
[2022-08-20 21:10:28.664676] attempting to download version 4/4: "!(\.HDR\.)" ... done
[2022-08-20 21:10:28.668713] [plex] item: "Death on the Nile" removed from Plex-Online-Library`s watchlist
[2022-08-20 21:10:28.893405] took 5.13s
[2022-08-20 21:10:28.893866] refreshing library section 1*

>

**Reply by itsToggle** [reply]
If you look closely at your version rule for 1080p, it seems that the is a space character at the end of the definition: "(1080p) " this will not match any release, since spaces are usually replaced by dots.

**Reply by itsToggle** [reply]
Though I'm not sure why the rest wasn't downloaded. I'll have a look.

**Reply by itsToggle** [reply]
Damn, found the error.. will update this immediately

**Reply by itsToggle** [reply]
fixed in latest commit. Make sure to change the 1080p rule as described earlier.

**Reply by Jauntiness** [reply]
Beautiful! Now all is working as intended on my end! Thanks for the quick fix 👍
